src/cs50/sql.py

```diff
@@ -357,6 +364,16 @@ def shutdown_session(exception=None):
                 elif command in ["DELETE", "UPDATE"]:
                     ret = result.rowcount
 
+                # If COMMIT or ROLLBACK, turn on autocommit mode
+                elif command in ["COMMIT", "ROLLBACK"] and "TO" not in statement:
+                    session.close()
+                    self._session = None
+
+                # If autocommit is on, commit and close
+                if self._session is None and command not in ["COMMIT", "ROLLBACK"]:
+                    session.commit()
+                    session.close()
```

Comment:
Instead of using the values of `session`, `self._session`, and `command` to determine whether or not we're in a transaction and based on that commit or not, would it perhaps be less confusing to use a `self._in_transaction` flag that's set on `BEGIN|START` and unset on `COMMIT|ROLLBACK`? Would that also allow us to reuse the same session across multiple `execute` calls without having to repeatedly close the session?

Reply:
Some explanation of my design decisions, though I'm happy to make any changes you'd prefer:

The `self._session` variable is how the transaction persists between calls to `execute` when the user wants to handle transactions themselves without autocommit. If we don't store the session then the transaction will either end up as a leaked database connection (if we don't close it) or will end up closed (making the user's subsequent queries worthless).

The reason the session is closed at the end of every `execute` call in autocommit is because often SQLAlchemy sessions have the same scope as transactions, so they're closed at the end of a transaction. This is just replicating that convention, but can be changed to keep the session open if desired -- the `__del__` method and Flask teardown function ensure the session will close eventually.

As for why I used `session` instead of `self._session` for execution, that's largely based on the existing codebase before these changes, where `connection` and `self._connection` are different things -- one stores the persistent connection, the other is used for execution in this specific call.

As for using a flag for `self._in_transaction`, that can certainly be done; I just thought it was repetitive and that adding an additional instance variable to the class was probably not the best design decision given the design decisions made by previous contributors (especially when a lot of the existing code does something very similar to checking if a variable is None, as by checking if the class hasattr). But this can obviously be added, and we could just use `self._session` (and even keep the session open), if preferred.

Reply:
> If we don't store the session then the transaction will either end up as a leaked database connection (if we don't close it) or will end up closed (making the user's subsequent queries worthless).

No no, I'm not saying we shouldn't store the session. We need to. I'm just saying that we could just rely on `self._session` instead of creating a separate `session` variable and then determining whether we're in a transaction or not by checking the values of `command` and/or `self._session` as we do on lines 364 and 368. So what I'm proposing is something like this:

```
if command is BEGIN or START:
  if self._in_transaction:
    raise exception

  set self._in_transaction

...

if self._session is None:
  set self._session to a new session

execute query

if not self._in_transaction:
  commit
elif command is COMMIT or ROLLBACK:
  commit
  unset self._in_transaction
  close session

# TODO how should executing COMMIT or ROLLBACK without a BEGIN or START behave?
```

Yes, this does add an additional instance variable, but I think it perhaps makes it easier to follow what we're trying to do? Oh, we're creating a new session if we don't have one and only committing if we're not a transaction or we're done with a transaction, instead of wondering, okay what's `session` and how's that different from `self._session`? And why does `session` sometimes have a value when `self._session` is `None`? etc.

Reply:
Ah! Got it. That makes sense. I'll modify to use the indicator. Last question is whether we want the session to persist between calls to `execute`? There's no reason it couldn't; in a regular script `__del__` will eventually close the session, and in Flask the request will close it too. And using the indicator variable eliminates the use of `self._session` as the way to figure out if we should autocommit.

Reply:
Given that `execute` could be called a few times within a single code block, and the applications that are meant to use the library are fairly simple, I would reuse the session across `execute` calls, but use a new session for every request (i.e., keep closing the session after request as we're doing now). Shouldn't be a big deal to keep the session open for a while as per https://docs.sqlalchemy.org/en/13/orm/session_transaction.html#managing-transactions, the session requests a connection upon executing a query, and keeps it open until a commit or a rollback is executed when it release the connection.

Reply:
As for how `COMMIT` or `ROLLBACK` without `BEGIN` or `START` behaves, right now there's no effect at all -- but for educational purposes it may be useful to throw an error explaining that transactions should be started with `BEGIN` or `START TRANSACTION`, e.g. `RuntimeError("transactions must be initiated with BEGIN or START TRANSACTION")` -- curious what @dmalan thinks.

Reply:
I would also try to run `COMMIT` or `ROLLBACK` without a `BEGIN` or `START` in MySQL and Postgres and see if that would throw an error or would just be a no-op and try to match this behavior, if `session.execute` doesn't handle that already (i.e., odds are if an error happens, `session.execute` itself would throw an exception).

Reply:
Just tested that. Only SQLite throws an error. I assume the other two are implicitly beginning new transactions, which I think is standard behavior based on readings from the past few days. From a pedagogical perspective, I think it may be worthwhile to throw our own error `if not self._in_transaction` -- especially if the user is trying to rollback and is surprised/confused that it doesn't work because they didn't realize they were in autocommit mode!

Reply:
Agreed, @jsarchibald!

Reply:
Ok, I'm moving the part of the code where that exception may be raised (also the part where we turn on autocommit) to _before_ the database execution, so as to preempt the SQLite default error -- Google says the answer to the default error is to **never** use `BEGIN`, which is advice directly opposed to what we're trying to convey.

Reply:
Have a URL of that thread?

Reply:
https://stackoverflow.com/questions/9773200/python-sqlite3-cannot-commit-no-transaction-is-active

Reply:
Can you recap how we're handling autocommit mode now? Are we disabling it when we see a `BEGIN` or `START TRANSACTION` in a student's argument to `execute`?

Reply:
The database-level autocommit is always disabled with SQLAlchemy, regardless of if the student is doing a transaction themselves. That's partly because we explicitly disable autocommit on the engine, and also because of the line that reads `dbapi_connection.isolation_level = None`. This was preexisting code designed to, in the words of the comment above it, `Disable underlying API's own emitting of BEGIN and COMMIT`. As for why SQLite throws an error here but PostgreSQL and MySQL don't, I'm not sure.

Our "autocommit" simply means our library commits at the end of the execution.

Fwiw, I think using a more descriptive error, like `transactions must be initiated with BEGIN or START TRANSACTION`, for all the databases (rather than relying on SQLite's own more difficult error) is probably just a better user experience.

Reply:
Sounds good! Ready to review PR when you are.